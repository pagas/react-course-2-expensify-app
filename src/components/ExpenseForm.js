import React from 'react';
import moment from 'moment';

import {SingleDatePicker} from "react-dates";


export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(()=>({error: 'Please provide description and amount!'}))
        } else {
            this.setState(()=>({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, // in cents format
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    onDescriptionChange = (e) => {
        this.setState(() => ({description: e.target.value}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        // only numbers, that might have 0-2 decimals
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }

    onNoteChange = (e) => {
        this.setState(() => ({note: e.target.value}))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}))
        }
    }

    onCalendarFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}))
    }

    render() {
        return <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                <input type="text"
                       name="description"
                       placeholder="Description"
                       autoFocus
                       value={this.state.description}
                       onChange={this.onDescriptionChange}
                />

                <input type="text"
                       name="amount"
                       placeholder="Amount"
                       value={this.state.amount}
                       onChange={this.onAmountChange}
                />

                <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={this.onDateChange} // PropTypes.func.isRequired
                    focused={this.state.calendarFocused} // PropTypes.bool
                    onFocusChange={this.onCalendarFocusChange} // PropTypes.func.isRequired
                    id="createdAt" // PropTypes.string.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />

                <textarea
                    placeholder="Add a note for your expense (optional)"
                    name="note"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                ></textarea>


                <button>Add expense</button>
            </form>
        </div>
    }
}


