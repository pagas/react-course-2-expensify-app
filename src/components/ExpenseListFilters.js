import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters'
import {DateRangePicker} from "react-dates";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null // can be startDate, endDate or null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() =>({calendarFocused}))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        if (e.target.value === 'amount') {
            this.props.sortByAmount();
        } else if (e.target.value === 'date'){
            this.props.sortByDate();
        }
    }

    render() {
        const {filters} = this.props;
        return (
            <div>
                <input type="text" value={filters.text} onChange={this.onTextChange}/>

                <select value={filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={filters.startDate} // momentPropTypes.momentObj or null,
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={filters.endDate} // momentPropTypes.momentObj or null,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }

}

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
})

const mapStateToProps = (state) => ({
    filters: state.filters
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
