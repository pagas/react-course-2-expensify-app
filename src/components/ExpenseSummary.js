import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import expenseTotal from '../selectors/expense-total';
import numeral from 'numeral';

export const ExpenseSummary = ({expenseCount = 0, expenseTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    return (
    <div>
        <h3>Viewing {expenseCount} {expenseWord} totalling {numeral(expenseTotal / 100).format('$0,0.00')}</h3>
    </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: expenseTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary);
