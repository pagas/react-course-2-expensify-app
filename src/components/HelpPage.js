import React from "react";
const HelpExpensePage = () => (
    <div>Help dashboard page</div>
)

export default HelpExpensePage;
