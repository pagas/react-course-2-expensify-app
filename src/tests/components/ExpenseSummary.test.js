import React from 'react';
import {ExpenseSummary} from "../../components/ExpenseSummary";
import {shallow} from 'enzyme';

test('Should show ExpenseSummary with count and total equal to 0', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={0} expenseTotal={0} />)
    expect(wrapper).toMatchSnapshot();
});

test('Should show empty ExpenseSummary with count equal to 1', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expenseTotal={233} />)
    expect(wrapper).toMatchSnapshot();
});

test('Should show correct ExpenseSummary with multiple count and totals', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={22} expenseTotal={23432} />)
    expect(wrapper).toMatchSnapshot();
});
