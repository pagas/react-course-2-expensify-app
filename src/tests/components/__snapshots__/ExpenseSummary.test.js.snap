// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should show ExpenseSummary with count and total equal to 0 1`] = `
<div>
  <h3>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h3>
</div>
`;

exports[`Should show correct ExpenseSummary with multiple count and totals 1`] = `
<div>
  <h3>
    Viewing 
    22
     
    expenses
     totalling 
    $234.32
  </h3>
</div>
`;

exports[`Should show empty ExpenseSummary with count equal to 1 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totalling 
    $2.33
  </h3>
</div>
`;
